### ISSUE TITLE
Flow thinks functions conform to Object "map" of {[string]: Object}

### ISSUE THREAD

**Issue by Benjamin-Dobell** [facebook/flow#6784]
Flow allows the following:

```javascript
// @flow
type SomeMap = {[string]: Object}

const okay : SomeMap = {hi: {}}
const huh : SomeMap = () => null
```

[Try Flow](https://flow.org/try/#0C4TwDgpgBAyg9gWwgWQIZigXigbwNoDOwATgJYB2A5gLoBcUA8gEYBWEAxsAL4BQP7cckShwA1qhBR68JGgzYcAC1L0cXXgKHAoigK6KpsRCnRYoACgCUWAHxRyugDaOeQA)

Functions will at the very least have a `name` property which is a `string` (https://www.ecma-international.org/ecma-262/9.0/index.html#sec-function-instances-name), thus they should be no more conforming than `{name: 'someString'}`, [which is correctly rejected](https://flow.org/try/#0C4TwDgpgBAyg9gWwgWQIZigXigbwNoDOwATgJYB2A5gLoBcUA8gEYBWEAxsAL4BQP7cckShwA1qhBR68JGgzYcAC1L0cXXgKHAoigK6KpsRCnRYoACgCUWAHxRyugDaO+m4cQhtOEACaGZJvK45KhI9ADk4VxQAPQxUAAqygRQpCkeXsC+ADRQTLraAAZ6ioVQqI4EcDqoKaj2odDAiqjA4XVQRGRU5eR+pG11znAA7r5QzcRwupSKAHQLQA).

**Comment by Benjamin-Dobell** [comment]
By the way, I suspect this is likely hiding quite a few bugs in people's code and libdefs. It's a fairly common pattern to offer an API that takes either an object map, or a function that when called is _supposed_ to return an object map e.g.

```javascript
// @flow
type Context = {version: number} // Some sort of context, can be anything
type Options = {[string]: Object}
type FunctionTakingContextReturningOptions = (Context) => Options

function config(options: Options | FunctionTakingContextReturningOptions) {
  if (typeof options === 'function') {
    options = options({version: 1})
  }

  /* Do stuff with options... */
}

config((arg1: string, arg2: number) : number => {
  console.log("What? You thought I'd take a Context and return Options?")
  console.log("Well, you be wrong... and you can't stop (validate) me. Muahahaha!")
  return 42 // Only the best numbers for you, fine sir.
})
```

[Try Flow](https://flow.org/try/#0C4TwDgpgBAwg9gO2BAHsKBeKBvAbhAJwGcBLRALigQFcBbAI0IF8oB6VqAZTluiLgLo4AMygBjRMjQAacQEMEURlAUhgACxIIA5gChQkKAHkwwMgiKYcAbSLACW7QF1KR+gCsIY4E33hoAGLUCN7mACpyANaO8EiowABKEMDUBAiOJmaIllgAFLFSwACUmAB8xqbmRLq6wsGhiOKIwiTauXCV2a6dFlAAPlBBIVkIEdE6BfFJKWkZPUQl2LpQUCSiuQYQIlAdIzkYWADkdcPmh4vLKzvzVrtVuXiEpBRQAIxMRZe+l6wAVFAAETgUDs1GEogA7iQNNc9gA6BFQX6sXTfCQIFptXJyAjaV6UOwOHSyHHaABMlBoDEIJUpdEYBDKOEu6P4ABsIHC2XA2gAiADq6jkwAA-FAAJpwahQDRS7TqdAASUOABMZVFoHJYJJ4ioEGqCMlUopMlURbzPitWXAOVyebkBRA2WzZCApUpoBCCIhtAi4Xq1W7pWIFId0HYOlBcrg5GySCrhRASrx-QBZahyIVZuQAQgtl0NM0UABYyWwOEYEGyQDL1NBGHYqPSnlBhAIoEHZC0EHwSAQ4aiikA)

**Comment by Benjamin-Dobell** [comment]
It seems the problem is a bit stranger than I realised:

__Passes Flow (Incorrect)__
```javascript
// @flow
type SomeMap = {[string]: {[string]: {[string]: {[string]: {[string]: Object}}}}}
const huh : SomeMap = () => null
```
[Try Flow](https://flow.org/try/#0C4TwDgpgBAyg9gWwgWQIZigXigbwNoDOwATgJYB2A5gLoBcuhJFN9+RZVdD7zXbTnegHkARgCsIAY2ABfOXIBQkuOSJQAFgFd1UevCRoM2ABQBKLAD4o5TQBtbCoA)

__Passes Flow (Incorrect)__
```javascript
// @flow
type SomeMap = {[string]: {[string]: {[string]: {[string]: {[string]: any}}}}}
const huh : SomeMap = () => null
```
[Try Flow](https://flow.org/try/#0C4TwDgpgBAyg9gWwgWQIZigXigbwNoDOwATgJYB2A5gLoBcuhJFN9+RZVdD7zXbTneqnIgAvuPEAoAMZxyRKAAsAroqj14SNBmwAKAJRYAfFHLKANuaA)

__Correctly catches the error:__
```javascript
// @flow
type SomeMap = {[string]: {[string]: {[string]: {[string]: {[string]: number}}}}}
const huh : SomeMap = () => null
```
[Try Flow](https://flow.org/try/#0C4TwDgpgBAyg9gWwgWQIZigXigbwNoDOwATgJYB2A5gLoBcuhJFN9+RZVdD7zXbTneuQCuCAEYRiAXxkyAUAGM45IlAAWwtVHrwkaDNgAUASiwA+KCIA2VoA)

That last one is particularly odd. It seems as though the issue isn't as simple as `Function` simply being internally represented as a callable `{[string]: Object}`, as I'd first thought, or else that last test ought to pass. Something else is going on.

**Comment by Benjamin-Dobell** [comment]
I'm not going to pretend to understand why this works, but I found a work-around:

```
// @flow
type SomeMap = {[string]: $Shape<Object>}

const okay : SomeMap = {hi: {}}
const huh : SomeMap = () => null
```

That will now correctly catch the error.

**Comment by wchargin** [comment]
Potentially related: <https://github.com/facebook/flow/issues/6490>

**Comment by gkz** [comment]
This is resolved as of 0.187
